src/maps/mod.rs

```diff
@@ -269,194 +237,187 @@ impl PerfMap {
-                        let header_bytes: [u8; size_of::<PerfEventHeader>()] = buf
-                            [..size_of::<PerfEventHeader>()]
-                            .try_into()
-                            .map_err(|_| OxidebpfError::BadPerfSample)?; // infallible, do not log
-                        std::ptr::read(header_bytes.as_ptr() as *const _)
-                    };
-
-                    let size = {
-                        let size_bytes: [u8; size_of::<u32>()] = buf[size_of::<PerfEventHeader>()
-                            ..(size_of::<u32>() + size_of::<PerfEventHeader>())]
-                            .try_into()
-                            .map_err(|_| OxidebpfError::BadPerfSample)?; // infallible, do not log
-                        u32::from_ne_bytes(size_bytes)
-                    };
-
-                    let data = {
-                        // drain the header + len fields; the rest is the data
-                        buf.drain(
-                            ..(std::mem::size_of::<PerfEventHeader>() + std::mem::size_of::<u32>()),
-                        )
-                        // might be unnecesary but I like this being explicit
-                        .for_each(|_| {});
-
-                        buf
-                    };
-
-                    let sample = PerfEventSample { header, size, data };
-                    Ok(Some(PerfEvent::Sample(sample)))
-                }
-                perf_event_type::PERF_RECORD_LOST => {
-                    let lost = &*(buf.as_ptr() as *const PerfEventLostSamples);
-                    Ok(Some(PerfEvent::Lost(lost.count)))
-                }
-                _ => Ok(None),
-            }
-        }
+                Ok(PerfMap {
+                    name: map_name.to_owned(),
+                    base_ptr: AtomicPtr::new(base_ptr),
+                    page_count,
+                    page_size,
+                    mmap_size,
+                    cpuid,
+                    ev_fd: fd,
+                    ev_name: "".to_owned(),
+                })
+            })
+            .collect()
     }
 
     /// Reads all available events
     ///
-    /// Stops reading either on a real error (and propagates it) or on
-    /// a OxidebpfError::NoPerfData and returns `None`.
+    /// Stops reading if it encounters an unexpected perf event.
+    ///
+    /// When the returned iterator is dropped it internally marks the
+    /// data as "read" so the ebpf program can re-use that
+    /// data. Because of this we should process the iterator fast as
+    /// to free space for more events.
+    ///
+    /// This function should actually be &mut self since it is
+    /// mutating data in the mmap (the tail to be exact) but that
+    /// makes it harder to use and it not used outside this
+    /// crate. Handle with care. Creating two iterators at the same
+    /// for one perfmap would create issues. TODO: change it to &mut
+    /// self or mark it as unsafe?
```

Comment:
I'd prefer to mark it as `unsafe`, in case we encounter subtle differences between the C contract and the Rust `&mut` contract.